reconcile/utils/smtp_client.py

```diff
@@ -3,24 +3,55 @@
 from email.mime.multipart import MIMEMultipart
 from email.mime.text import MIMEText
 from email.utils import formataddr
-from typing import Any, Iterable, Union
+from typing import Iterable, Optional
+from pydantic import BaseModel
 
 from sretoolbox.utils import retry
 
-from reconcile.utils.secret_reader import SecretReader
+from reconcile.utils.secret_reader import SecretReader, SupportsSecret
 
 
-class SmtpClient:
-    def __init__(self, settings: dict[str, Any]) -> None:
-        smtp_config = self.get_smtp_config(settings)
-        self.host: str = smtp_config["server"]
-        self.port: int = int(smtp_config["port"])
-        self.user: str = smtp_config["username"]
-        self.passwd: str = smtp_config["password"]
-        self.mail_address: str = settings["smtp"]["mailAddress"]
-        self.timeout: int = settings["smtp"].get("timeout", 30)
+class SmtpCredentials(BaseModel):
+    server: str
+    port: int
+    username: str
+    password: str
+    # require_tls: bool - currently not in use
+
 
-        self._client: Union[smtplib.SMTP, None] = None
+def get_smtp_credentials(
+    secret_reader: SecretReader, secret: SupportsSecret
+) -> SmtpCredentials:
+    """Retrieve SMTP credentials from config or vault."""
+    # This will change later when SecretReader fully supports 'SupportsSecret'
+    data = secret_reader.read_all(
+        {
+            "path": secret.path,
+            "field": secret.field,
+            "format": secret.q_format,
+            "version": secret.version,
+        }
+    )
+    return SmtpCredentials(**data)
+
+
+class SmtpClient:
+    def __init__(
+        self,
+        host: str,
+        port: int,
+        username: str,
+        password: str,
+        mail_address: str,
+        timeout: int = 30,
+    ) -> None:
+        self.host = host
```

Comment:
nit: this could go either way ultimately. I think if we have many args (think 6+) then it does start to make sense to have dataclass(es) to combine the configs that are logically related. In this case you already have `SmtpCredentials`, so is there any reason not to just make `__init__`:

```python
def __init__(mail_address: str, credentials: SmtpCredentials, timeout: int = 30)
```

I don't think there's anything wrong with it as-is other than that we go through the trouble of building the object in all of the usage, and don't use it:

```python
        smtp_credentials = get_smtp_credentials(
            secret_reader=SecretReader(settings=queries.get_secret_reader_settings()),
            secret=smtp_settings.credentials,
        )
        smtp_client = SmtpClient(
            host=smtp_credentials.server,
            port=smtp_credentials.port,
            username=smtp_credentials.username,
            password=smtp_credentials.password,
            mail_address=smtp_settings.mail_address,
            timeout=smtp_settings.timeout or 30,
        )
```

When I used this example originally, I was thinking that we'd have a parameter per setting if we weren't going to bother adding a data class for this use case (because it really could go either way).